baseapp/baseapp.go

```diff
@@ -708,17 +685,14 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 		}
 	}
 
-	// Create a new Context based off of the existing Context with a MultiStore branch
-	// in case message processing fails. At this point, the MultiStore
-	// is a branch of a branch.
-	runMsgCtx, msCache := app.cacheTxContext(ctx, txBytes)
-
 	// Attempt to execute all messages and only update state if all messages pass
 	// and we're in DeliverTx. Note, runMsgs will never return a reference to a
 	// Result if any single message fails or does not have a registered Handler.
-	result, err = app.runMsgs(runMsgCtx, msgs, mode)
-
-	if err == nil {
+	_ = ctx.RunAtomic(func(runMsgCtx sdk.Context) error {
+		result, err = app.runMsgs(runMsgCtx, msgs, mode)
+		if err != nil {
+			return err
+		}
```

Comment:
it's not "happens to", but on purpose, because it need to write to the outer `result`, so I make it write to the `err` as well.

Reply:
Yes, but this behavior is surprising, subtle, and fragile. The atomic func should report errors as returned error values, and its caller should inspect those returned errors rather than ignoring them as it does now.

Reply:
make sense, did a small change, should make it clearer.